internal/provider/resources/resource_team_roles.go

```diff
@@ -0,0 +1,329 @@
+	if diagnostic != nil {
+		diags.Append(diagnostic)
+		return diags
+	}
+
+	diags = data.ReadFromResponse(ctx, teamId, teamRoles.JSON200)
+	if diags.HasError() {
+		return diags
+	}
+
+	return nil
+}
+
+func (r *teamRolesResource) Create(
+	ctx context.Context,
+	req resource.CreateRequest,
+	resp *resource.CreateResponse,
+) {
+	var data models.TeamRoles
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	diags := r.MutateRoles(ctx, &data)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	// Save data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *teamRolesResource) Read(
+	ctx context.Context,
+	req resource.ReadRequest,
+	resp *resource.ReadResponse,
+) {
+	var data models.TeamRoles
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	teamId := data.TeamId.ValueString()
+
+	// get request
+	teamRoles, err := r.iamClient.GetTeamWithResponse(
+		ctx,
+		r.organizationId,
+		teamId,
+	)
+	if err != nil {
+		tflog.Error(ctx, "failed to get team_roles", map[string]interface{}{"error": err})
+		resp.Diagnostics.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to get team_roles, got error: %s", err),
+		)
+		return
+	}
+	statusCode, diagnostic := clients.NormalizeAPIError(ctx, teamRoles.HTTPResponse, teamRoles.Body)
+	// If the resource no longer exists, it is recommended to ignore the errors
+	// and call RemoveResource to remove the resource from the state. The next Terraform plan will recreate the resource.
+	if statusCode == http.StatusNotFound {
+		resp.State.RemoveResource(ctx)
+		return
+	}
+	if diagnostic != nil {
+		resp.Diagnostics.Append(diagnostic)
+		return
+	}
+
+	// Generate subjectRoles from the get team API response
+	subjectRoles := iam.SubjectRoles{
+		OrganizationRole: lo.ToPtr(iam.SubjectRolesOrganizationRole(teamRoles.JSON200.OrganizationRole)),
+		WorkspaceRoles:   teamRoles.JSON200.WorkspaceRoles,
+		DeploymentRoles:  teamRoles.JSON200.DeploymentRoles,
+	}
+	diags := data.ReadFromResponse(ctx, teamId, &subjectRoles)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	tflog.Trace(ctx, fmt.Sprintf("read a team_roles resource: %v", teamId))
+
+	// Save data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *teamRolesResource) Update(
+	ctx context.Context,
+	req resource.UpdateRequest,
+	resp *resource.UpdateResponse,
+) {
+	var data models.TeamRoles
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	diags := r.MutateRoles(ctx, &data)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
```

Comment:
could consider adding tflog.Trace logging here?

```suggestion
tflog.Trace(ctx, fmt.Sprintf("updated a team_roles resource for team '%v'", data.TeamId.ValueString()))
```